beam/core/src/main/java/cz/o2/proxima/beam/direct/io/OffsetRestrictionTracker.java

```diff
@@ -0,0 +1,164 @@
+     */
+    public static StartEndOffset initialRestriction() {
+      return new StartEndOffset();
+    }
+
+    /** @return restriction that reads from given offset (exclusive) */
+    public static StartEndOffset startingFrom(Offset start) {
+      return new StartEndOffset(start, true);
+    }
+
+    @Getter @Nullable private final Offset startOffset;
+    @Getter private final boolean startInclusive;
+    @Getter @Nullable private volatile Offset endOffsetInclusive;
+    // weather to disallow adding claim past endOffset
+    @Getter private volatile boolean extensible;
+    private transient boolean finished = false;
+
+    private StartEndOffset(@Nullable Offset start, boolean inclusive) {
+      this.startOffset = start;
+      this.startInclusive = inclusive;
+      extensible = true;
+    }
+
+    private StartEndOffset() {
+      this(null, false);
+    }
+
+    private StartEndOffset(Offset startExclusive) {
+      this(Objects.requireNonNull(startExclusive), false);
+    }
+
+    public StartEndOffset(Offset startOffsetExclusive, Offset endOffsetInclusive) {
+      this.startOffset = Objects.requireNonNull(startOffsetExclusive);
+      this.startInclusive = false;
+      this.endOffsetInclusive = Objects.requireNonNull(endOffsetInclusive);
+      this.extensible = false;
+    }
+
+    boolean getAndCommitClaim(Offset offset) {
+      if (!extensible) {
+        if (finished) {
+          return false;
+        }
+        finished = offset.equals(endOffsetInclusive);
+        return true;
+      }
+      endOffsetInclusive = offset;
+      return true;
+    }
+
+    @Nullable
+    public StartEndOffset asPrimary() {
+      if (endOffsetInclusive == null) {
+        return null;
+      }
+      return new StartEndOffset(startOffset, endOffsetInclusive);
+    }
+
+    public StartEndOffset asResidual() {
+      return new StartEndOffset(endOffsetInclusive);
+    }
+
+    public boolean isInitial() {
+      return startOffset == endOffsetInclusive && endOffsetInclusive == null;
+    }
+
+    void terminate() {
+      this.extensible = false;
+      this.finished = true;
+    }
+
+    @Override
+    public OffsetRestrictionTracker newTracker() {
+      Preconditions.checkState(
+          !finished, "Tracker can be created only for not-yet split restrictions");
+      Preconditions.checkState(
+          extensible, "Tracker can be created only for not-yet split restrictions");
+      return new OffsetRestrictionTracker(this);
+    }
+  }
+
+  private final StartEndOffset currentRestriction;
+
+  private OffsetRestrictionTracker(StartEndOffset initial) {
+    this.currentRestriction = initial;
+  }
+
+  @Override
+  public boolean tryClaim(Offset position) {
+    return currentRestriction.getAndCommitClaim(position);
+  }
+
+  @Override
+  public StartEndOffset currentRestriction() {
+    return currentRestriction;
+  }
+
+  @Override
+  public @Nullable SplitResult<StartEndOffset> trySplit(double fractionOfRemainder) {
```

Comment:
btw from our discussion earlier -> `   * <p>{@code fractionOfRemainder = 0} means a checkpoint is required.`

fractionOfRemainder=0 is also passed when we return ProcessCountinuation.resume()

Reply:
There is (currently) no other value passed to this call and it is ignored.